pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,276 @@
+package ftp
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/jlaffaye/ftp"
+)
+
+// Conn struct embeds the *ftp.ServerConn returned by ftp server on successful connection.
+type Conn struct {
+	*ftp.ServerConn
+}
+
+// Retr wraps the ftp retrieve method to return a ftpResponse interface type.
+func (c *Conn) Retr(path string) (ftpResponse, error) {
+	return c.ServerConn.Retr(path)
+}
+
+func (c *Conn) RetrFrom(path string, offset uint64) (ftpResponse, error) {
+	return c.ServerConn.RetrFrom(path, offset)
+}
+
+var (
+	errEmptyFilename           = errors.New("empty filename")
+	errEmptyPath               = errors.New("empty path")
+	errEmptyDirectory          = errors.New("empty directory")
+	errInvalidArg              = errors.New("invalid filename/directory")
+	directoryAlreadyExistError = "550 Create directory operation failed."
+	transferComplete           = "226 Transfer complete."
+)
+
+// ftpFileSystem represents a file system interface over FTP.
+type ftpFileSystem struct {
+	*ftpFile
+	conn   ServerConn
+	config *Config
+	logger Logger
+}
+
+// Config represents the FTP configuration.
+type Config struct {
+	Host      string // FTP server hostname
+	User      string // FTP username
+	Password  string // FTP password
+	Port      int    // FTP port
+	RemoteDir string // Remote directory path. Base Path for all FTP Operations.
+}
+
+// New initializes a new instance of ftpFileSystem with provided configuration.
+func New(config *Config) FileSystem {
+	return &ftpFileSystem{config: config}
+}
+
+// UseLogger sets the logger interface for the FTP file system.
+func (f *ftpFileSystem) UseLogger(logger interface{}) {
+	if l, ok := logger.(Logger); ok {
+		f.logger = l
+	}
+}
+
+// UseMetrics sets the metrics for the ftpFileSystem client which asserts the Metrics interface.
+// Currently not implemented.
+func (*ftpFileSystem) UseMetrics(_ interface{}) {
+
+}
+
+// Connect establishes a connection to the FTP server and logs in.
+func (f *ftpFileSystem) Connect() {
+	ftpServer := fmt.Sprintf("%v:%v", f.config.Host, f.config.Port)
+
+	const dialTimeout = 5 * time.Second
+
+	conn, err := ftp.Dial(ftpServer, ftp.DialWithTimeout(dialTimeout))
+	if err != nil {
+		f.logger.Errorf("Failed to connect to FTP server: %v", err)
+		return
+	}
+
+	f.conn = &Conn{conn}
+
+	err = conn.Login(f.config.User, f.config.Password)
+	if err != nil {
+		f.logger.Errorf("Failed to login: %v", err)
+	}
```

Comment:
Missing `return`? All it does is log a line after this, but it would still be confusing to see in the logs:
```
Failed to login: foo error
Login Successful
```

Reply:
Noted.